cpp/src/text/replace.cu

```diff
@@ -0,0 +1,164 @@
+ * are left unchanged.
+ *
+ * This should be called first to compute the size of each output string and then a second
+ * time to fill in the allocated output buffer for each string.
+ */
+struct replace_tokens_fn {
+  cudf::column_device_view const d_strings;  ///< strings to tokenize
+  cudf::column_device_view const d_targets;  ///< strings to replace
+  cudf::column_device_view const d_repls;    ///< replacement strings
+  cudf::string_view const d_delimiter;       ///< delimiter characters for tokenizing
+  const int32_t* d_offsets{};
+  char* d_chars{};
+
+  __device__ cudf::size_type operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) return 0;
+    cudf::string_view const d_str = d_strings.element<cudf::string_view>(idx);
+    cudf::size_type nbytes        = d_str.size_bytes();
+    auto in_ptr                   = d_str.data();
+    char* out_ptr                 = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    cudf::size_type last_pos      = 0;
+    characters_tokenizer tokenizer(d_str, d_delimiter);
+    while (tokenizer.next_token()) {
+      auto token_pos = tokenizer.token_byte_positions();
+      cudf::string_view token{d_str.data() + token_pos.first, (token_pos.second - token_pos.first)};
+      // check if the token matches any of the targets
+      for (auto tidx = 0; tidx < d_targets.size(); ++tidx) {
+        cudf::string_view d_target = d_targets.element<cudf::string_view>(tidx);
+        if (token.compare(d_target) == 0) {
+          // match found
+          // retrieve the corresponding replacement string or
+          // if only one repl string, use that one for all targets
+          cudf::string_view d_repl = d_repls.size() == 1 ? d_repls.element<cudf::string_view>(0)
+                                                         : d_repls.element<cudf::string_view>(tidx);
+          nbytes += d_repl.size_bytes() - token.size_bytes();  // total output bytes
+          if (out_ptr) {
+            // copy over string upto the token location
+            out_ptr = cudf::strings::detail::copy_and_increment(
+              out_ptr, in_ptr + last_pos, token_pos.first - last_pos);
+            // copy over replacement string
+            out_ptr  = cudf::strings::detail::copy_string(out_ptr, d_repl);
+            last_pos = token_pos.second;  // update last byte position for this string
+          }
+          break;
+        }
+      }
+    }
+    // copy the remainder of the string bytes to the output buffer
+    if (out_ptr) memcpy(out_ptr, in_ptr + last_pos, d_str.size_bytes() - last_pos);
+    return nbytes;
+  }
+};
+
+}  // namespace
+
+// detail APIs
+
+// zero or more character tokenizer
+std::unique_ptr<cudf::column> replace_tokens(cudf::strings_column_view const& strings,
+                                             cudf::strings_column_view const& targets,
+                                             cudf::strings_column_view const& repls,
```

Comment:
Is `repls` short for `replacement_strings`? For me, REPL = [read eval print loop](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop) and although I am sure most people won't be confused in this context, abbreviating this parameter name and not the others seems odd. I would vote to extend the parameter name to `replacements`.

Reply:
Here is the etymology of `repl`
This came from me extending from the Pandas space. This parameter would match up to some Python/Cython API and I wanted to be able to follow a parameter all the way down without changing its name in each layer.

I used `repl` in the case of https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.slice_replace.html#pandas.Series.str.slice_replace

```
Series.str.slice_replace(self, start=None, stop=None, repl=None)
  Parameters:
     start - Left index position to use for the slice.
     stop  - Right index position to use for the slice.
     repl  - String for replacement
```

Since the target user of these APIs would be Data Scientists (most likely familiar with Pandas), I wanted the parameter to be more familiar/recognizable to that space.

Next, I felt it would be good to consistently name parameters [across all (strings) APIs that had that same purpose](https://docs.rapids.ai/api/libcudf/nightly/group__strings__replace.html) and that is why `repl` is used here in this API as well.

- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace.hpp#L65
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace.hpp#L104
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace.hpp#L150
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace.hpp#L172
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace_re.hpp#L48
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace_re.hpp#L69
- https://github.com/rapidsai/cudf/blob/efebcd1452692ee15f3d30627a9ef3d0cafa85d5/cpp/include/cudf/strings/replace_re.hpp#L89

I may have missed some.

Naming things is the hardest part of this job :-)